src/ATen/templates/StorageDerived.cpp

```diff
@@ -12,12 +12,42 @@ namespace at {
 ${Storage}::${Storage}(Context* context, std::size_t storage_size)
   : storage(${THStorage}_newWithSize(${state,} storage_size)), context(context) {}
 
+#if ${isCUDA}
+static cudaError_t call_deleter(void * ctx, void * data) {
+  auto fnptr = (std::function<void(void*)>*) ctx;
+  (*fnptr)(data);
+  delete fnptr;
+  return cudaSuccess;
+}
+static THCDeviceAllocator storage_deleter = {
+  nullptr,
+  nullptr,
+  call_deleter,
+  nullptr,
+  nullptr,
+};
+#else
+static void call_deleter(void * ctx, void * data) {
+  auto fnptr = (std::function<void(void*)>*) ctx;
+  (*fnptr)(data);
+  delete fnptr;
+}
+static THAllocator storage_deleter = {
+  nullptr,
+  nullptr,
+  call_deleter,
+};
+#endif
+
 ${Storage}::${Storage}(Context* context,
-  void * data, std::size_t size)
-  : storage(${THStorage}_newWithData(${state,} static_cast<${THScalarType}*>(data), size)),
+  void * data, std::size_t size, const std::function<void(void*)> & deleter)
```

Comment:
here

Reply:
I didn't add a default here because the derived storages are a private API. I believe all the places in the public API where deleters are used take defaults.